### ISSUE TITLE
omniperf analyze statistics does not match understanding

### ISSUE THREAD

**Issue by nishshah0** [AMDResearch/omniperf#66]
I have using using omniperf to analyze some of the applications. I ran a simple 8x8x8 gemm in BF16 data format using following command line
`omniperf profile -n gemm_m8_k8_n8] -d 1 --device 0 -- ./rocBLAS/build/release/clients/staging/rocblas-bench -m 8 -k 8 -n 8 -f gemm_ex -r bf16_r --compute_type f32_r -i 1 -j 1  --device 0`

after running `omniperf analyze -p gemm_m8_k8_n8` I get following output
![image](https://user-images.githubusercontent.com/122632692/212390735-e14f7518-7ded-430b-9fb0-8fb4e7fe16b5.png)

The highlighted metric MFMA Flops (BF16) does not make sense. I expect 8x8x8x2 = 1024 flops.

Kernel takes 14.8 us, look below
![image](https://user-images.githubusercontent.com/122632692/212391348-293001da-5198-489d-9fbe-42db5d81342d.png)

So I expect 1024/(14.8 * 1e-6) = 69.2 Million Flops ~ 0.068 GFLOPS. 

But I see 4.4 Gflops. How is this calculated?

**Comment by coleramos425** [comment]
MFMA FLOPs (BF16) leverages `SQ_INSTS_VALU_MFMA_MOPS_BF16`. Its value is computed using 

```
AVG((SQ_INSTS_VALU_MFMA_MOPS_BF16 * 512) / (EndNs - BeginNs))
```

@arghdos may be able to provide a more detailed description of the counter

**Comment by arghdos** [comment]
The 512 BF16 flops/instruction value in the MI-200 equation appears to be incorrect, at least according to: https://gpuopen.com/learn/amd-lab-notes/amd-lab-notes-matrix-cores-readme/, but we'll need to double check.

That said @shaw586, assuming you did a single BF16 operation is probably incorrect.
You need to take the value of `MFMA-BF16` in the `MFMA Arithmetic Instr Mix` and multiply that by the number of waves launched to get the total number of BF16 operations.
Then take the total number of BF16 ops, and multiply by the FLOP/OP count to get total FLOPS (then normalized by time).

@coleramos425 -- I noticed a separate issue where the values in the above section can only be normalized by the # of waves:

https://github.com/AMDResearch/omniperf/blob/main/src/omniperf_analyze/configs/gfx90a/1000_compute-unit-instruction-mix.yaml#L171

Can you open something to track?

**Comment by shaw586** [comment]
This the instruction mix statistics
![image](https://user-images.githubusercontent.com/122632692/213012416-01388518-bb68-4db4-a500-23ee5c15185a.png)

This the wavefront statistics
![image](https://user-images.githubusercontent.com/122632692/213012231-83386d8f-7591-4b21-8944-8a2f64e92175.png)

What metric do I multiply MFMA-BF16 with from wavefront statistics?

**Comment by arghdos** [comment]
2 MFMA-BF16 ops / wave * 4 wavefronts * 1024 FLOPs / BF16 op = 8192 BF16 FLOPs

**Comment by shaw586** [comment]
8192/14880 = 0.55 Gflops. The speed of light show 4.4 Gflops. Which seams exactly 8x higher. How is That metric in speed of light calculated?

**Comment by arghdos** [comment]
>How is That metric in speed of light calculated?

https://github.com/AMDResearch/omniperf/blob/main/src/omniperf_analyze/configs/gfx90a/0200_system-speed-of-light.yaml#L45

Not sure what's happening here, if you can share the `pmc_perf.csv` file that's being generated in `workloads/<project name>/mi200` perhaps @coleramos425  can take a look

**Comment by shaw586** [comment]
Attaching the pmc_perf.csv
[pmc_perf.csv](https://github.com/AMDResearch/omniperf/files/10440330/pmc_perf.csv)

This is the command I used to run 
`omniperf profile -n gemm_m8_k8_n8] -d 1 --device 0 -- ./rocBLAS/build/release/clients/staging/rocblas-bench -m 8 -k 8 -n 8 -f gemm_ex -r bf16_r --compute_type f32_r -i 1 -j 1  --device 0`

**Comment by coleramos425** [comment]
This dispatch has a duration of **14880** ns (i.e. EndNs-BeginNs) and a `SQ_INSTS_VALU_MFMA_MOPS_BF16` of **128** 

(SQ_INSTS_VALU_MFMA_MOPS_BF16 * 512) / (EndNs - BeginNs) =
128 * 512 / 14880 = 65536 / 14880 = 4.4 Gflops

@xiaomin-amd claims the arithmetic is correct. I'll forward to let him weigh in.

**Comment by xiaomin-amd** [comment]
[AMD Official Use Only - General]

Please dump the matching disassembly for analysis.

**Comment by shaw586** [comment]
Can someone please educate me how to dump disassembly? I am using omniprofile with rocblas-bench.

Also its really surprising to see a 8x8x8 gemm executing 64k flops. Thats 64x higher than what I am specifiying!

**Comment by coleramos425** [comment]
@shaw586 seems to be several methods but `objdump -d /path/to/binary` is a straightforward option included in GNU development tools

**Comment by arghdos** [comment]
[AMD Official Use Only - General]

https://github.com/ROCm-Developer-Tools/HIP/blob/develop/docs/markdown/obj_tooling.md
[https://opengraph.githubassets.com/266d044dfc3c35230e218935f1d79cea9ee350507c5876bfbe9af616cad16f90/ROCm-Developer-Tools/HIP]<https://github.com/ROCm-Developer-Tools/HIP/blob/develop/docs/markdown/obj_tooling.md>
HIP/obj_tooling.md at develop · ROCm-Developer-Tools/HIP<https://github.com/ROCm-Developer-Tools/HIP/blob/develop/docs/markdown/obj_tooling.md>
ROCm Code Object tooling roc-obj Examples: Extract all ROCm code objects from a list of executables Extract all ROCm code objects from a list of executables, and disassemble them Extract all ROCm code objects from a list of executables into dir/ Extract only ROCm code objects matching regex over Target ID Low-Level Tooling URI syntax: List available ROCm Code Objects: rocm-obj-ls Extract ROCm ...
github.com

________________________________

**Comment by shaw586** [comment]
I am sorry, that readme specifies how I can generate objdump of compiled application where I have executable. But how do I use this with rocblas-bench?

**Comment by arghdos** [comment]
It sorta depends on how rocblas-bench is set up.  If the kernels are compiled into rocblas-bench, you should be able to do a roc-obj -d /path/to/rocblas/bench to get the ISA for all the kernels there.

It is possible they're loaded at runtime though from another library, in which case you can use the same command on that library, or the HIP environment variable [`export GPU_DUMP_CODE_OBJECT=1`](https://github.com/ROCm-Developer-Tools/ROCclr/blob/main/utils/flags.hpp#L212) to dump the kernels you use at runtime.

**Comment by shaw586** [comment]
I following following steps,
`export GPU_DUMP_CODE_OBJECT=1
./rocBLAS/build/release/clients/staging/rocblas-bench -m 8 -k 8 -n 8 -f gemm_ex -r bf16_r --compute_type f32_r -i 1 -j 1 --device 0`

This resulted in generation of bunch of object files.
![image](https://user-images.githubusercontent.com/122632692/214445505-13a8aa4a-66e5-4af3-a7b8-44e14ca90f6a.png)

When I open them with command
`roc-obj -d _code_object0000.o`

I get following error
`Error: No kernel section found
error: no executables specified`

**Comment by shaw586** [comment]
ping! Any further directions?

**Comment by coleramos425** [comment]
Not sure how much help I can provide outside of pointing you to the README for roc-obj
https://github.com/ROCm-Developer-Tools/HIP/blob/develop/docs/markdown/obj_tooling.md

Looks to me like `-d` is supposed to take an executable, no?

**Comment by coleramos425** [comment]
@shaw586 please update the thread if you are still actively attempting to generate a dump. Otherwise, I will archive this issue.

Thank you.

**Comment by shaw586** [comment]
Yes, I am trying to generate the dump. Been bouncing different places to try and get some help. Havent found my luck yet. Currently waiting on rocBLAS team to respond. If you can help find someone who knows how to dump assembly while using rocBLAS, I would appreciate it!

**Comment by shaw586** [comment]
Here is the assembly dump of the kernel being executed, 
[TensileLibrary_Type_BB_HPA_Contraction_l_Ailk_Bljk_Cijk_Dijk_gfx90a_m8_k8_n8_objdump.zip](https://github.com/AMDResearch/omniperf/files/10914120/TensileLibrary_Type_BB_HPA_Contraction_l_Ailk_Bljk_Cijk_Dijk_gfx90a_m8_k8_n8_objdump.zip)

**Comment by coleramos425** [comment]
Closing due to inactivity. Please re-open if you still have questions.
